agent/engine/host_resource_manager.go

```diff
@@ -16,20 +16,226 @@
 package engine
 
 import (
+	"fmt"
+	"sync"
+
 	"github.com/aws/amazon-ecs-agent/agent/ecs_client/model/ecs"
 	"github.com/aws/amazon-ecs-agent/agent/logger"
 	"github.com/aws/amazon-ecs-agent/agent/utils"
 )
 
+// TODO remove this once resource, consume are used
+//lint:file-ignore U1000 Ignore all unused code
+
 // initialHostResource keeps account of each task in
 type HostResourceManager struct {
-	initialHostResource map[string]*ecs.Resource
-	consumedResource    map[string]*ecs.Resource
+	initialHostResource       map[string]*ecs.Resource
+	consumedResource          map[string]*ecs.Resource
```

Comment:
A bit of a meta question here. IIUC the keys to these two maps are resource types such as  CPU, memory, and ports. Is the full range of resource types managed by `HostResourceManager` type static or dynamic? If it is static, then have we thought about adding dedicated fields for each resource type and using new easy-to-use types? `ecs.Resource` type seems quite error prone as it is not well typed.

Reply:
There will be 5 resource types accounted for in HostResourceManager currently (cpu, memory, ports_tcp, ports_udp, gpu). We are choose `ecs.Resource` type as this is the same format agent uses to report host resources to ecs when registering instance ([here](https://github.com/aws/amazon-ecs-agent/blob/883211e24d9255023a68f6351001c291af8a699b/agent/api/ecsclient/client.go#L188)).

Reply:
I see. IIUC the goal of `HostResourceManager` is to manage the state of resources and `ecs.Resource` type doesn't seem to be easy-to-use for that purpose. It is probably designed for communicating the resources state to the backend. 

Also since each resource has its own mutation function (CPU is a continuous variable from 0 to 100 while ports are a set), it might be worthwhile to define an interface for resource and handle the state changes behind in the implementation.

This is a non-blocking comment and I am just sharing my opinion. :)

Reply:
I agree a separate type for `HostResourceManager` can be a bit more useful. On the other hand I also see that by re-using `ecs.Resource` we are able to piggy back on an existing implementation which could be easy in it's own way as it is declared to define ecs resources (albeit to communicate to backend, perhaps). And using a map, accessing values becomes fairly easy.

If it is not much of a hassle, I will stay with `ecs.Resource` implementation, as previous, parallel (and future) work is already using this type. But I will be willing to change this upon recommendation :)